compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/cfa/ControlFlowInfos.kt

```diff
@@ -84,11 +84,126 @@ class PropertyInitializationInfo(
     override val constructor: (PersistentMap<FirPropertySymbol, EventOccurrencesRange>) -> PropertyInitializationInfo =
         ::PropertyInitializationInfo
 
-    override fun put(key: FirPropertySymbol, value: EventOccurrencesRange): PropertyInitializationInfo {
-        return super.put(key, value) as PropertyInitializationInfo
+    override val empty: () -> PropertyInitializationInfo =
+        ::EMPTY
+
+    override fun merge(other: PropertyInitializationInfo): PropertyInitializationInfo {
+        return merge(this, other)
+    }
+}
+
+sealed class PathAwareControlFlowInfo<S : ControlFlowInfo<S, *, *>>(
```

Comment:
I think it's good to add another generic self-type `P : PathAwareControlFlowInfo<S, P>` (or something like that), to avoid `super.someCall(...) as SpecificImplementation` in all implementations

Reply:
Thanks! I wanted to do that, but somehow didn't find a good way to do so, now I understand why `ControlFlowInfo` has that self-type parameter!